CMakeLists.txt

```diff
@@ -217,18 +217,22 @@ endif()
 # extra needed libraries
 if(WIN32)
   list(APPEND mi_libraries psapi shell32 user32 advapi32 bcrypt)
+  set(pc_libraries "-lpsapi -lshell32 -luser32 -ladvapi32 -lbcrypt")
 else()
   find_library(MI_LIBPTHREAD pthread)
   if (MI_LIBPTHREAD)                      
     list(APPEND mi_libraries ${MI_LIBPTHREAD})
+    set(pc_libraries "-pthread")
```

Comment:
- should this be "-lpthread" ? maybe not?
- maybe always use `set(pc_libraries "${pc_libraries} <lib>"` ??
- can we maybe use the "mi_libraries" variable directly to generate the `pc_libraries` -- for example, by replacing all `-l` by a space?  That would avoid all the duplication

Reply:
- No, because -lpthread is a linker option only, whereas `-pthread` is documented as:
  - a preprocessor option: "Define additional macros required for using the POSIX threads library.  You should use this option consistently for both compilation and linking."
  - a linker option: "Link with the POSIX threads library. [...] On some targets this option also sets flags for the preprocessor, so it should be used consistently for both compilation and linking."
- Is this a style suggestion? pc_libraries does not exist yet, so it seemed shorter to not mention it, but I can change that if you want
- Unfortunately we cannot use mi_libraries, because cmake treats these differently (specially). If we used mi_libaries directly in the pkg-config file, we would get: `Libs.private: /usr/lib/libpthread.a;/usr/lib/librt.a;atomic`; this would be broken for two reasons
  - because semicolons are CMake's non-type-safe way of using 1980s technology to implement bourne shell lookalike lists (bourne shell coding style frequently used lists as space-delimited strings, because there wasn't a list type; likewise, cmake does not have a list type, but chose the semicolon and implemented semicolon-delimited strings for accumulating values). You cannot pass a semicolon to a compiler, and cmake's configure_file doesn't automatically replace semicolons with spaces the way Make/ninja file generation does when generating a command line
  - `atomic` is not a valid linker command line option, but cmake allows passing that as one anyway, and before producing the Makefiles or ninja files, it will first check to see whether `atomic` is the name of an IMPORTED or ALIAS cmake target, and if so use that, otherwise check if it is an absolute path, and use that as is, and otherwise it will add `-l` to the beginning when running the compiler/linker

  Furthermore, pkg-config is supposed to have `-lfoo`, never `/path/to/libfoo.so`, and some people critically rely on that.

Reply:
Aside: I really wish cmake had a native pkg-config generator. In [Meson](https://mesonbuild.com/) this is very easy:

```meson
mi_deps = [
    dependency('threads'),
    cc.find_library('rt'),
    cc.find_library('atomic'),
]
mimalloc_lib = library('mimalloc', ..., dependencies: mi_deps)

# automatically adds all dependencies as their preferred representation
# for:
#     found libraries, this is `-lfoo`
#     threads, this is `-pthread`
#     found dependencies, this is the `Requires.private:` field listing the *.pc file to merge in
pkgconfig.generate(mimalloc_lib)
```

Reply:
ah, ok, those are all good reasons. I wish indeed cmake did more -- I am trying to battle the expanding complexity of the CMakeFiles but to no avail :-)